### PR TITLE
Fix GA loss for Deepspeed

### PR DESCRIPTION
**Pull request by timjeffrey10**
# What does this PR do?

This PR addresses an issue where the `grad_norm` differs when the `gradient_accumulation_steps` changes when using DeepSpeed. The issue does not occur when using DDP.

In the accelerate library, the `Accelerator` class ignores the `gradient_accumulation_steps` parameter when using DeepSpeed because DeepSpeed handles loss scaling internally. The relevant code in `Accelerate` is:
[Reference](https://github.com/huggingface/accelerate/blob/78b8126bff9882a66074a082cb2a80aa89026118/src/accelerate/accelerator.py#L2234-L2236)
```python
        if self.distributed_type != DistributedType.DEEPSPEED:
            # deepspeed handles loss scaling by gradient_accumulation_steps in its `backward`
            loss = loss / self.gradient_accumulation_steps
```
In DeepSpeed, loss scaling occurs in the `_scale_loss_by_gas` method when the `backward` function is called. While setting `gradient_accumulation_steps` to 1 in the `Accelerator` disables loss scaling, it cannot be set to 1 in DeepSpeed. This is because DeepSpeed relies on this parameter to manage internal operations like counting `micro_steps` in its `step` function.

In DeepSpeed `backward` function, it allow us to pass `scale_wrt_gas` to turn off the loss scaling. The `_scale_loss_by_gas` will not be called.
[Reference](https://github.com/microsoft/DeepSpeed/blob/de4596bedc61100db9385b5d99efd9025db13a7d/deepspeed/runtime/engine.py#L2028-L2029)
```python
        if do_gradient_reduction and self.gradient_accumulation_steps() > 1 and scale_wrt_gas:
            loss = self._scale_loss_by_gas(loss.float())
```

This PR passes the `kwargs["scale_wrt_gas"]=False` to the `backward` function when the DeepSpeed is enabled to prevent potential loss scaling in DeepSpeed engine.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @sunmarc @muellerzr

**Review by muellerzr — COMMENTED** [review]
Hi! When submitting changes to the GA fixes now, we're going to ask that you provide training logs (wandb) of a baseline vs GA-fixed model to make absolutely sure that all goes as intended. Also please run the related GA slow tests and let us know if they pass in `test_trainer.py`

**Comment by timjeffrey10** [comment]
Thanks for the review.
I have committed another solution to pass `scale_wrt_gas` to the backward function, which prevents the loss scaling in DeepSpeed engine.
Another solution is to set `scale_wrt_gas` after initializing DeepSpeed engine (i.e., after calling `accelerator.prepare`). I have no idea how to set it during initilizing :(
```python
        model, self.optimizer = self.accelerator.prepare(self.model, self.optimizer)
        if self.is_deepspeed_enabled:
            model.scale_wrt_gas = False
```
This can help not to always add `scale_wrt_gas` to kwargs in `training_step`. If reviewer prefer this implementation, I can try to modify it.

I will upload the wandb logs and testing result ASAP.

**Comment by timjeffrey10** [comment]
# Training logs (wandb)
- DeepSpeed: ZeRO-2
- 2 GPUs, Global batch size = 512
- Compare `gradient_accumulation_steps`=1 and 256

Here is the orinal version, the `grad_norm` is extreme small when `gradient_accumulation_steps`=256
image

Here is the revised version, the `grad_norm` is almost the same in both `gradient_accumulation_steps`=1 and 256
image

# GA Tests Results
Below is the screenshot of the results of releted GA tests I found. Please let me know if any tests are missing.
image

@muellerzr If there are any problem, please let me know. Thank a lot.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3714,6 +3714,10 @@ def training_step(
             if not self.model_accepts_loss_kwargs and self.compute_loss_func is None:
                 loss = loss / self.args.gradient_accumulation_steps
 
+            # Turning off loss scaling w.r.t. gradient accumulation when DeepSpeed is enabled
```
Comment (muellerzr) [inline src/transformers/trainer.py:3717]: Let's also link to this PR, since it includes W&B logs

**Review by muellerzr — APPROVED** [review]
Beautiful! Thanks for adding the logs there :hugs: 

One request just to make a comment linking to this PR :)

I'm okay with this way of doing it

cc @SunMarc

File: src/transformers/trainer.py (hunk 1 above)
Comment (timjeffrey10) [reply src/transformers/trainer.py:3717]: Done. Thank you~

**Review by ArthurZucker — APPROVED** [review]
Nice, thanks for the quick update
